File: test/spec/server-discovery-and-monitoring/errors/generate-error-tests.py
import itertools
import os
import subprocess
import sys

# Require Python 3.7+ for ordered dictionaries so that the order of the
# generated tests remain the same.
if sys.version_info[:2] < (3, 7):
    print('ERROR: This script requires Python >= 3.7, not:')
    print(sys.version)
    print('Usage: python3 %s' % (sys.argv[0]))
    exit(1)


dirname = os.path.dirname
DIR = dirname(os.path.realpath(__file__))
SOURCE = dirname(dirname(dirname(DIR)))


def template(filename):
    fullpath = os.path.join(DIR, filename)
    with open(fullpath, 'r') as f:
        return f.read()


def write_test(filename, data):
    fullpath = os.path.join(DIR, filename + '.yml')
    with open(fullpath, 'w') as f:
        f.write(data)

    print(f"Generated {fullpath}")


# Maps from error_name to (error_code,)
ERR_CODES = {
    'InterruptedAtShutdown': (11600,),
    'InterruptedDueToReplStateChange': (11602,),
    'NotPrimaryOrSecondary': (13436,),
    'PrimarySteppedDown': (189,),
    'ShutdownInProgress': (91,),
    'NotWritablePrimary': (10107,),
    'NotPrimaryNoSecondaryOk': (13435,),
    'LegacyNotPrimary': (10058,),
}


def create_stale_tests():
    tmp = template('stale-topologyVersion.yml.template')
    for error_name in ERR_CODES:
        test_name = f'stale-topologyVersion-{error_name}'
        error_code, = ERR_CODES[error_name]
        data = tmp.format(**locals())
        write_test(test_name, data)

TV_GREATER = '''
      topologyVersion:
        processId:
          "$oid": '000000000000000000000001'
        counter:
          "$numberLong": "2"'''
TV_GREATER_FINAL =  '''
          processId:
            "$oid": '000000000000000000000001'
          counter:
            "$numberLong": "2"'''
TV_CHANGED = '''
      topologyVersion:
        processId:
          "$oid": '000000000000000000000002'
        counter:
          "$numberLong": "1"'''
TV_CHANGED_FINAL =  '''
          processId:
            "$oid": '000000000000000000000002'
          counter:
            "$numberLong": "1"'''

# Maps non-stale error description to:
# (error_topology_version, final_topology_version)
NON_STALE_CASES = {
    'topologyVersion missing': ('', ' null'),
    'topologyVersion greater': (TV_GREATER, TV_GREATER_FINAL),
    'topologyVersion proccessId changed': (TV_CHANGED, TV_CHANGED_FINAL),
}


def create_non_stale_tests():
    tmp = template('non-stale-topologyVersion.yml.template')
    for error_name, description in itertools.product(
            ERR_CODES, NON_STALE_CASES):
        test_name = f'non-stale-{description.replace(" ", "-")}-{error_name}'
        error_code, = ERR_CODES[error_name]
        error_topology_version, final_topology_version = NON_STALE_CASES[description]
        # On 4.2+, only ShutdownInProgress and InterruptedAtShutdown will
        # clear the pool.
        if error_name in ("ShutdownInProgress", "InterruptedAtShutdown"):
            final_pool_generation = 1
        else:
            final_pool_generation = 0

        data = tmp.format(**locals())
        write_test(test_name, data)


WHEN = ['beforeHandshakeCompletes', 'afterHandshakeCompletes']
STALE_GENERATION_COMMAND_ERROR = '''
    type: command
    response:
      ok: 0
      errmsg: {error_name}
      code: {error_code}
      topologyVersion:
        processId:
          "$oid": '000000000000000000000001'
        counter:
          "$numberLong": "2"'''
STALE_GENERATION_NETWORK_ERROR = '''
    type: {network_error_type}'''


def create_stale_generation_tests():
    tmp = template('stale-generation.yml.template')
    # Stale command errors
    for error_name, when in itertools.product(ERR_CODES, WHEN):
        test_name = f'stale-generation-{when}-{error_name}'
        error_code, = ERR_CODES[error_name]
        stale_error = STALE_GENERATION_COMMAND_ERROR.format(**locals())
        data = tmp.format(**locals())
        write_test(test_name, data)
    # Stale network errors
    for network_error_type, when in itertools.product(
            ['network', 'timeout'], WHEN):
        error_name = network_error_type
        test_name = f'stale-generation-{when}-{network_error_type}'
        stale_error = STALE_GENERATION_NETWORK_ERROR.format(**locals())
        data = tmp.format(**locals())
        write_test(test_name, data)


def create_pre_42_tests():
    tmp = template('pre-42.yml.template')
    # All "not writable primary"/"node is recovering" clear the pool on <4.2
    for error_name in ERR_CODES:
        test_name = f'pre-42-{error_name}'
        error_code, = ERR_CODES[error_name]
        data = tmp.format(**locals())
        write_test(test_name, data)


def create_post_42_tests():
    tmp = template('post-42.yml.template')
    for error_name in ERR_CODES:
        test_name = f'post-42-{error_name}'
        error_code, = ERR_CODES[error_name]
        # On 4.2+, only ShutdownInProgress and InterruptedAtShutdown will
        # clear the pool.
        if error_name in ("ShutdownInProgress", "InterruptedAtShutdown"):
            final_pool_generation = 1
        else:
            final_pool_generation = 0
        data = tmp.format(**locals())
        write_test(test_name, data)


create_stale_tests()
create_non_stale_tests()
create_stale_generation_tests()
create_pre_42_tests()
create_post_42_tests()

print('Running make')
subprocess.run(f'cd {SOURCE} && make', shell=True, check=True)

File: test/unit/spec/sdam-errors/generate-error-tests.ts
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import * as path from 'path';

// TypeScript port of the former generate-error-tests.py. These SDAM application-error
// tests are internally maintained (not synced from the specifications repo), so they live
// here under test/unit/spec/sdam-errors rather than in the vendored test/spec tree.
//
// Run with: npx ts-node test/unit/spec/sdam-errors/generate-error-tests.ts
//
// Unlike the Python version (which emitted YAML and relied on the specifications repo's
// `make` to convert to JSON), this script writes the final .json fixtures directly.

const DIR = __dirname;

const template = (filename: string): string => fs.readFileSync(path.join(DIR, filename), 'utf8');

/**
 * Re-implementation of Python's `str.format(**vars)` for named fields. The YAML templates
 * only use simple `{field}` placeholders (no literal `{{`/`}}` escapes), so a single pass
 * over named fields is sufficient.
 */
function format(tmpl: string, vars: Record<string, string | number>): string {
  return tmpl.replace(/\{(\w+)\}/g, (_match, name: string) => {
    if (!(name in vars)) {
      throw new Error(`template referenced unknown field: {${name}}`);
    }
    return String(vars[name]);
  });
}

function writeTest(filename: string, data: string): void {
  const fullpath = path.join(DIR, `${filename}.json`);
  const parsed = yaml.load(data);
  fs.writeFileSync(fullpath, `${JSON.stringify(parsed, null, 2)}\n`);
  console.log(`Generated ${fullpath}`);
}

// Maps from error_name to error_code
const ERR_CODES: Record<string, number> = {
  InterruptedAtShutdown: 11600,
  InterruptedDueToReplStateChange: 11602,
  NotPrimaryOrSecondary: 13436,
  PrimarySteppedDown: 189,
  ShutdownInProgress: 91,
  NotWritablePrimary: 10107,
  NotPrimaryNoSecondaryOk: 13435,
  LegacyNotPrimary: 10058
};

// On 4.2+, only ShutdownInProgress and InterruptedAtShutdown clear the pool.
const clearsPool = (errorName: string): 0 | 1 =>
  errorName === 'ShutdownInProgress' || errorName === 'InterruptedAtShutdown' ? 1 : 0;

function createStaleTests(): void {
  const tmp = template('stale-topologyVersion.yml.template');
  for (const [errorName, errorCode] of Object.entries(ERR_CODES)) {
    const testName = `stale-topologyVersion-${errorName}`;
    writeTest(testName, format(tmp, { error_name: errorName, error_code: errorCode }));
  }
}

const TV_GREATER = `
      topologyVersion:
        processId:
          "$oid": '000000000000000000000001'
        counter:
          "$numberLong": "2"`;
const TV_GREATER_FINAL = `
          processId:
            "$oid": '000000000000000000000001'
          counter:
            "$numberLong": "2"`;
const TV_CHANGED = `
      topologyVersion:
        processId:
          "$oid": '000000000000000000000002'
        counter:
          "$numberLong": "1"`;
const TV_CHANGED_FINAL = `
          processId:
            "$oid": '000000000000000000000002'
          counter:
            "$numberLong": "1"`;

// Maps non-stale error description to [error_topology_version, final_topology_version]
const NON_STALE_CASES: Record<string, [string, string]> = {
  'topologyVersion missing': ['', ' null'],
  'topologyVersion greater': [TV_GREATER, TV_GREATER_FINAL],
  'topologyVersion proccessId changed': [TV_CHANGED, TV_CHANGED_FINAL]
};

function createNonStaleTests(): void {
  const tmp = template('non-stale-topologyVersion.yml.template');
  for (const [errorName, errorCode] of Object.entries(ERR_CODES)) {
    for (const [description, [errorTV, finalTV]] of Object.entries(NON_STALE_CASES)) {
      const testName = `non-stale-${description.replace(/ /g, '-')}-${errorName}`;
      writeTest(
        testName,
        format(tmp, {
          error_name: errorName,
          error_code: errorCode,
          error_topology_version: errorTV,
          final_topology_version: finalTV,
          final_pool_generation: clearsPool(errorName)
        })
      );
    }
  }
}

const WHEN = ['beforeHandshakeCompletes', 'afterHandshakeCompletes'] as const;
const STALE_GENERATION_COMMAND_ERROR = `
    type: command
    response:
      ok: 0
      errmsg: {error_name}
      code: {error_code}
      topologyVersion:
        processId:
          "$oid": '000000000000000000000001'
        counter:
          "$numberLong": "2"`;
const STALE_GENERATION_NETWORK_ERROR = `
    type: {network_error_type}`;

function createStaleGenerationTests(): void {
  const tmp = template('stale-generation.yml.template');
  // Stale command errors
  for (const [errorName, errorCode] of Object.entries(ERR_CODES)) {
    for (const when of WHEN) {
      const testName = `stale-generation-${when}-${errorName}`;
      const staleError = format(STALE_GENERATION_COMMAND_ERROR, {
        error_name: errorName,
        error_code: errorCode
      });
      writeTest(
        testName,
        format(tmp, {
          error_name: errorName,
          error_code: errorCode,
          when,
          stale_error: staleError
        })
      );
    }
  }
  // Stale network errors
  for (const networkErrorType of ['network', 'timeout']) {
    for (const when of WHEN) {
      const testName = `stale-generation-${when}-${networkErrorType}`;
      const staleError = format(STALE_GENERATION_NETWORK_ERROR, {
        network_error_type: networkErrorType
      });
      writeTest(
        testName,
        format(tmp, {
          error_name: networkErrorType,
          network_error_type: networkErrorType,
          when,
          stale_error: staleError
        })
      );
    }
  }
}

function createPost42Tests(): void {
  const tmp = template('post-42.yml.template');
  for (const [errorName, errorCode] of Object.entries(ERR_CODES)) {
    writeTest(
      `post-42-${errorName}`,
      format(tmp, {
        error_name: errorName,
        error_code: errorCode,
        final_pool_generation: clearsPool(errorName)
      })
    );
  }
}

// NOTE: createPre42Tests (pre-42.yml.template) is intentionally not invoked. Those fixtures
// target maxWireVersion 8 (MongoDB 4.2), which is below the driver's minimum supported wire
// version, so they were pruned. The template is retained for history.

createStaleTests();
createNonStaleTests();
createStaleGenerationTests();
createPost42Tests();
